File: Leetcode/PY/Leetcode1609-奇偶树.py
class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        q = []
        q.append(root)
        st = True
        while len(q):
            n,pre,cur = len(q),0,0
            for i in range(n):
                t = q.pop(0)
                pre,cur = cur,t.val
                if t.left: q.append(t.left)
                if t.right: q.append(t.right)
                if st:
                    if cur % 2 == 0:return False
                    elif i > 0 and pre >= cur: return False
                else:
                    if cur % 2 == 1: return False
                    elif i > 0 and pre <= cur: return False
            st = not st
        return True 

File: Leetcode/JS/Leetcode1609-奇偶树.js
var isEvenOddTree = function(root) {
    const q = [];
    let st = true;
    q.push(root);
    while(q.length){
        const n = q.length;
        let pre = -1,cur = 0;
        for(let i = 0;i < n;i++){
            let t = q.shift();
            pre = cur,cur = t.val;
            if(t.left) q.push(t.left);
            if(t.right) q.push(t.right);
            if(st){
                if(cur % 2 == 0) return false;
                if(i > 0 && pre >= cur) return false;
            }else{
                if(cur % 2 == 1) return false;
                if(i > 0 && pre <= cur) return false;
            }
        }
        st = !st;
    }
    return true;
};